egs/wsj/s5/steps/nnet3/libs/xconfig_layers.py

```diff
@@ -0,0 +1,679 @@
+    # other layers, and get dimensions from them.
+    def ConvertToDescriptor(self, descriptor_string, all_layers):
+        prev_names = xconfig_utils.GetPrevNames(all_layers, self)
+        tokens = xconfig_utils.TokenizeDescriptor(descriptor_string, prev_names)
+        pos = 0
+        (descriptor, pos) = xconfig_utils.ParseNewDescriptor(tokens, pos, prev_names)
+        # note: 'pos' should point to the 'end of string' marker
+        # that terminates 'tokens'.
+        if pos != len(tokens) - 1:
+            raise RuntimeError("Parsing Descriptor, saw junk at end: " +
+                            ' '.join(tokens[pos:-1]))
+        return descriptor
+
+    # Returns the dimension of a Descriptor object.
+    # This is a convenience function used in SetConfigs.
+    def GetDimForDescriptor(self, descriptor, all_layers):
+        layer_to_dim_func = lambda name: xconfig_utils.GetDimFromLayerName(all_layers, self, name)
+        return descriptor.Dim(layer_to_dim_func)
+
+    # Returns the 'final' string form of a Descriptor object, as could be used
+    # in config files.
+    # This is a convenience function provided for use in child classes;
+    def GetStringForDescriptor(self, descriptor, all_layers):
+        layer_to_string_func = lambda name: xconfig_utils.GetStringFromLayerName(all_layers, self, name)
+        return descriptor.ConfigString(layer_to_string_func)
+
+    # Name() returns the name of this layer, e.g. 'affine1'.  It does not
+    # necessarily correspond to a component name.
+    def Name(self):
+        return self.name
+
+    ######  Functions that might be overridden by the child class: #####
+
+    # child classes should override this.
+    def SetDefaultConfigs(self):
+        raise RuntimeError("Child classes must override SetDefaultConfigs().")
+
+    # child classes should override this.
+    def CheckConfigs(self):
+        pass
+
+    # This function, which may be (but usually will not have to be) overrideden
+    # by child classes, returns a list of keys/names of config variables that
+    # will be interpreted as Descriptors.  It is used in the function
+    # 'NormalizeDescriptors()'.  This implementation will work
+    # layer types whose only Descriptor-valued config is 'input'.
+
+    # If a child class adds more config variables that are interpreted as
+    # descriptors (e.g. to read auxiliary inputs), or does not have an input
+    # (e.g. the XconfigInputLayer), it should override this function's
+    # implementation to something like: `return ['input', 'input2']`
+    def GetDescriptorConfigs(self):
+        return [ 'input' ]
+
+    # Returns a list of all qualifiers (meaning auxiliary outputs) that this
+    # layer supports.  These are either 'None' for the regular output, or a
+    # string (e.g. 'projection' or 'memory_cell') for any auxiliary outputs that
+    # the layer might provide.  Most layer types will not need to override this.
+    def Qualifiers(self):
```

Comment:
I am a bit confused about the choice of the name Qualifiers for auxiliary outputs. Could you please give a brief explanation.